pkg/querier/frontend/frontend_querier_queues.go

```diff
@@ -0,0 +1,227 @@
+		uq.queriers = shuffleQueriersForUser(uq.seed, maxQueriers, q.sortedQueriers, nil)
+	}
+
+	return uq.ch
+}
+
+// Finds next queue for the querier. To support fair scheduling between users, client is expected
+// to pass last user index returned by this function as argument. Is there was no previous
+// last user index, use -1.
+func (q *queues) getNextQueueForQuerier(lastUserIndex int, querier string) (chan *request, string, int) {
+	uid := lastUserIndex
+
+	for iters := 0; iters < len(q.users); iters++ {
+		uid = uid + 1
+
+		// Don't use "mod len(q.users)", as that could skip users at the beginning of the list
+		// for example when q.users has shrunk since last call.
+		if uid >= len(q.users) {
+			uid = 0
+		}
+
+		u := q.users[uid]
+		if u == "" {
+			continue
+		}
+
+		q := q.userQueues[u]
+
+		if q.queriers != nil {
+			if _, ok := q.queriers[querier]; !ok {
+				// This querier is not handling the user.
+				continue
+			}
+		}
+
+		return q.ch, u, uid
+	}
+	return nil, "", uid
+}
+
+func (q *queues) addQuerierConnection(querier string) {
+	conns := q.querierConnections[querier]
+
+	q.querierConnections[querier] = conns + 1
+
+	// First connection from this querier.
+	if conns == 0 {
+		q.sortedQueriers = append(q.sortedQueriers, querier)
+		sort.Strings(q.sortedQueriers)
+
+		q.recomputeUserQueriers()
+	}
+}
+
+func (q *queues) removeQuerierConnection(querier string) {
+	conns := q.querierConnections[querier]
+	if conns <= 0 {
+		return
+	}
+
+	conns--
+	if conns > 0 {
+		q.querierConnections[querier] = conns
+	} else {
+		delete(q.querierConnections, querier)
+
+		ix := sort.SearchStrings(q.sortedQueriers, querier)
+		q.sortedQueriers = append(q.sortedQueriers[:ix], q.sortedQueriers[ix+1:]...)
+
+		q.recomputeUserQueriers()
+	}
+}
+
+func (q *queues) recomputeUserQueriers() {
+	scratchpad := make([]string, 0, len(q.sortedQueriers))
+
+	for _, uq := range q.userQueues {
+		uq.queriers = shuffleQueriersForUser(uq.seed, uq.maxQueriers, q.sortedQueriers, scratchpad)
+	}
+}
+
+// Scratchpad is used for shuffling, to avoid new allocations. If nil, new slice is allocated.
+// shuffleQueriersForUser returns nil if queriersToSelect is 0 or there are not enough queriers to select from.
+// In that case *all* queriers should be used.
+func shuffleQueriersForUser(userSeed int64, queriersToSelect int, allSortedQueriers []string, scratchpad []string) map[string]struct{} {
+	if queriersToSelect == 0 || len(allSortedQueriers) <= queriersToSelect {
+		return nil
+	}
+
+	result := make(map[string]struct{}, queriersToSelect)
+	rnd := rand.New(rand.NewSource(userSeed))
+
+	scratchpad = scratchpad[:0]
+	scratchpad = append(scratchpad, allSortedQueriers...)
+
+	last := len(scratchpad) - 1
+	for i := 0; i < queriersToSelect; i++ {
+		r := rnd.Intn(last + 1)
+		result[scratchpad[r]] = struct{}{}
+		// move selected item to the end, it won't be selected anymore.
+		scratchpad[r], scratchpad[last] = scratchpad[last], scratchpad[r]
+		last--
+	}
```

Comment:
Is it preferable consistent hashing here? I know it's not strictly required, but I'm wondering if it's preferable. I believe this algorithm doesn't guarantee it because of "swap selected item to the end", which basically shuffle the slice if we add or remove a querier to the pool.

Reply:
Probably we don't care about consistent hashing, but I would like to have your opinion too.

Reply:
Consistent as in when new querier is added, it doesn't totally change sharding for all users.

We currently don't do that and agreed that it's ok in this case. We may want to revisit that decision at some point.

Reply:
Ok to keep it as it. I believe consistent hashing is not a required property here and we can always revisit it.